src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c

```diff
@@ -453,27 +463,48 @@ zocl_xclbin_load_pdi(struct drm_zocl_dev *zdev, void *data)
 	}
 
 	mutex_lock(&zdev->zdev_xclbin_lock);
-	/* Check unique ID */
-	if (zocl_xclbin_same_uuid(zdev, &axlf_head->m_header.uuid)) {
-		DRM_INFO("%s The XCLBIN already loaded, uuid: %pUb",
-			 __func__, &axlf_head->m_header.uuid);
-		mutex_unlock(&zdev->zdev_xclbin_lock);
-		return ret;
-	}
-
-	write_lock(&zdev->attr_rwlock);
 
-	/* Get full axlf header */
 	size_of_header = sizeof(struct axlf_section_header);
 	num_of_sections = axlf_head->m_header.m_numSections-1;
 	xclbin = (char __user *)axlf;
-	ret =
-	    !ZOCL_ACCESS_OK(VERIFY_READ, xclbin, axlf_head->m_header.m_length);
-	if (ret) {
-		ret = -EFAULT;
+
+	ret = !ZOCL_ACCESS_OK(VERIFY_READ, xclbin, axlf_head->m_header.m_length);
+        if (ret) {
+                ret = -EFAULT;
+                goto out;
+        }
+
+	write_lock(&zdev->attr_rwlock);
+
+	zocl_read_sect(AIE_RESOURCES, &aie_res, axlf, xclbin);
+
+	/* Check unique ID */
+	if (zocl_xclbin_same_uuid(zdev, &axlf_head->m_header.uuid)) {
+		if (is_aie_only(axlf)) {
+                        ret = zocl_load_aie_only_pdi(zdev, axlf, xclbin, NULL);
+                        if (ret)
+                                DRM_WARN("read xclbin: fail to load AIE");
+                        else
+                                zocl_create_aie(zdev, axlf, aie_res);
+		} else {
+			DRM_INFO("%s The XCLBIN already loaded, uuid: %pUb",
+				 __func__, &axlf_head->m_header.uuid);
+		}
 		goto out;
 	}
 
+	if (is_aie_only(axlf)) {
+                ret = zocl_load_aie_only_pdi(zdev, axlf, xclbin, NULL);
+                if (ret)
+                        goto out;
+        }
+
+	size = zocl_read_sect(AIE_METADATA, &zdev->aie_data.data, axlf, xclbin);
+	if (size > 0 ) {
+		zdev->aie_data.size = size;
+		zocl_create_aie(zdev, axlf, aie_res);
+	}
+
 	size = zocl_offsetof_sect(BITSTREAM_PARTIAL_PDI, &section_buffer,
```

Comment:
seems you can move this 
if (aie_only)
   read AIE, then load AIE into a function

Thus we can avoid dup code.

Reply:
@xdavidz made changes as suggested

Reply:
Hi @rbramand-xilinx , i don't see this change yet. Can you please make this change to avoid dup code.

Reply:
@chvamshi-xilinx made changes as requested